File: Motoshift/lib/models/usuario.dart
// Mapeado para a entidade `Usuario` no Spring Boot / MySQL
// Tabela: usuarios
class Usuario {
  final int? id;
  final String nome;
  final String email;
  final String telefone;
  final TipoUsuario tipo; // 'LOJISTA' | 'MOTOBOY'
  final String? documentoFederal; // CNPJ (lojista) ou CNH (motoboy)
  final String? fotoPerfil;
  final double score;
  final double? mediaAvaliacao;

  // Dados pessoais
  final DateTime? dataNascimento;
  final String? cidade;
  final String? estado;

  // CNH e Veículo (motoboy)
  final String? cnhNumero;
  final String? cnhCategoria;
  final DateTime? cnhValidade;
  final String? veiculoModelo;
  final String? veiculoPlaca;
  final int? veiculoAno;
  final String? veiculoCor;

  // Lojista
  final String? nomeFantasia;
  final String? enderecoComercial;

  final DateTime? criadoEm;

  const Usuario({
    this.id,
    required this.nome,
    required this.email,
    required this.telefone,
    required this.tipo,
    this.documentoFederal,
    this.fotoPerfil,
    this.score = 5.0,
    this.mediaAvaliacao,
    this.dataNascimento,
    this.cidade,
    this.estado,
    this.cnhNumero,
    this.cnhCategoria,
    this.cnhValidade,
    this.veiculoModelo,
    this.veiculoPlaca,
    this.veiculoAno,
    this.veiculoCor,
    this.nomeFantasia,
    this.enderecoComercial,
    this.criadoEm,
  });

  factory Usuario.fromJson(Map<String, dynamic> json) {
    DateTime? parseDate(String key) {
      final v = json[key];
      if (v is String && v.isNotEmpty) return DateTime.parse(v);
      return null;
    }

    return Usuario(
      id: json['id'] as int?,
      nome: json['nome'] as String,
      email: json['email'] as String,
      telefone: json['telefone'] as String,
      tipo: TipoUsuario.values.byName((json['tipo'] as String).toLowerCase()),
      documentoFederal: json['documentoFederal'] as String?,
      fotoPerfil: json['fotoPerfil'] as String?,
      score: (json['score'] as num?)?.toDouble() ?? 5.0,
      mediaAvaliacao: (json['mediaAvaliacao'] as num?)?.toDouble(),
      dataNascimento: parseDate('dataNascimento'),
      cidade: json['cidade'] as String?,
      estado: json['estado'] as String?,
      cnhNumero: json['cnhNumero'] as String?,
      cnhCategoria: json['cnhCategoria'] as String?,
      cnhValidade: parseDate('cnhValidade'),
      veiculoModelo: json['veiculoModelo'] as String?,
      veiculoPlaca: json['veiculoPlaca'] as String?,
      veiculoAno: json['veiculoAno'] as int?,
      veiculoCor: json['veiculoCor'] as String?,
      nomeFantasia: json['nomeFantasia'] as String?,
      enderecoComercial: json['enderecoComercial'] as String?,
      criadoEm: parseDate('criadoEm'),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      if (id != null) 'id': id,
      'nome': nome,
      'email': email,
      'telefone': telefone,
      'tipo': tipo.name,
      if (documentoFederal != null) 'documentoFederal': documentoFederal,
      if (fotoPerfil != null) 'fotoPerfil': fotoPerfil,
    };
  }

  Usuario copyWith({
    int? id,
    String? nome,
    String? email,
    String? telefone,
    TipoUsuario? tipo,
    String? documentoFederal,
    String? fotoPerfil,
    double? score,
    double? mediaAvaliacao,
    DateTime? dataNascimento,
    String? cidade,
    String? estado,
    String? cnhNumero,
    String? cnhCategoria,
    DateTime? cnhValidade,
    String? veiculoModelo,
    String? veiculoPlaca,
    int? veiculoAno,
    String? veiculoCor,
    String? nomeFantasia,
    String? enderecoComercial,
    DateTime? criadoEm,
  }) {
    return Usuario(
      id: id ?? this.id,
      nome: nome ?? this.nome,
      email: email ?? this.email,
      telefone: telefone ?? this.telefone,
      tipo: tipo ?? this.tipo,
      documentoFederal: documentoFederal ?? this.documentoFederal,
      fotoPerfil: fotoPerfil ?? this.fotoPerfil,
      score: score ?? this.score,
      mediaAvaliacao: mediaAvaliacao ?? this.mediaAvaliacao,
      dataNascimento: dataNascimento ?? this.dataNascimento,
      cidade: cidade ?? this.cidade,
      estado: estado ?? this.estado,
      cnhNumero: cnhNumero ?? this.cnhNumero,
      cnhCategoria: cnhCategoria ?? this.cnhCategoria,
      cnhValidade: cnhValidade ?? this.cnhValidade,
      veiculoModelo: veiculoModelo ?? this.veiculoModelo,
      veiculoPlaca: veiculoPlaca ?? this.veiculoPlaca,
      veiculoAno: veiculoAno ?? this.veiculoAno,
      veiculoCor: veiculoCor ?? this.veiculoCor,
      nomeFantasia: nomeFantasia ?? this.nomeFantasia,
      enderecoComercial: enderecoComercial ?? this.enderecoComercial,
      criadoEm: criadoEm ?? this.criadoEm,
    );
  }
}

enum TipoUsuario { lojista, motoboy }

File: backend/src/main/java/com/motoshift/entity/Usuario.java
package com.motoshift.entity;

import jakarta.persistence.*;
import org.hibernate.annotations.DynamicUpdate;

import java.time.LocalDate;
import java.time.LocalDateTime;

/**
 * Conta da plataforma — lojista ou entregador.
 *
 * {@code @DynamicUpdate} porque esta linha tem dois escritores que não se
 * conhecem: a edição de perfil (que salva a entidade inteira) e o controle de
 * tentativas de login (que só mexe em dois campos, por UPDATE direto). Com o
 * UPDATE de todas as colunas, que é o padrão do Hibernate, salvar o perfil logo
 * depois de um erro de senha regravaria o contador com o valor antigo.
 */
@Entity
@DynamicUpdate
@Table(name = "usuarios")
public class Usuario {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false)
    private String nome;

    @Column(nullable = false, unique = true)
    private String email;

    @Column(nullable = false)
    private String telefone;

    // "lojista" | "motoboy"
    @Column(nullable = false)
    private String tipo;

    private String documentoFederal;

    private String fotoPerfil;

    // Hash BCrypt, sempre. Cadastro e massa de demonstração passam pelo
    // PasswordEncoder, e a V9 converteu as contas antigas que ainda estavam em
    // texto puro — o login não aceita outro formato.
    @Column(nullable = false)
    private String senha;

    @Column(nullable = false)
    private Double score = 5.0;

    private Double mediaAvaliacao;

    // ── Dados pessoais ─────────────────────────────────────────
    private LocalDate dataNascimento;
    private String cidade;
    private String estado;

    // ── CNH e Veículo (motoboy) ────────────────────────────────
    private String cnhNumero;
    private String cnhCategoria;          // "A" | "AB"
    private LocalDate cnhValidade;

    private String veiculoModelo;         // ex: "Honda CG 160"
    private String veiculoPlaca;          // ex: "ABC-1D23"
    private Integer veiculoAno;
    private String veiculoCor;

    // ── Lojista ────────────────────────────────────────────────
    private String nomeFantasia;
    private String enderecoComercial;

    // ── RF01: bloqueio por tentativas de login ─────────────────
    // No banco, e não num mapa em memória: o bloqueio precisa sobreviver a um
    // redeploy, valer igual em todas as instâncias e só existir para contas que
    // existem (o mapa era indexado pelo e-mail digitado e crescia com qualquer
    // e-mail inventado). Escritos só pelos UPDATEs do UsuarioRepository.
    @Column(nullable = false)
    private Integer tentativasLogin = 0;

    private LocalDateTime bloqueadoAte;

    @Column(nullable = false, updatable = false)
    private LocalDateTime criadoEm;

    @PrePersist
    private void prePersist() {
        criadoEm = LocalDateTime.now();
        if (score == null) score = 5.0;
        if (tentativasLogin == null) tentativasLogin = 0;
    }

    // --- Getters e Setters ---

    public Long getId() { return id; }

    public String getNome() { return nome; }
    public void setNome(String nome) { this.nome = nome; }

    public String getEmail() { return email; }
    public void setEmail(String email) { this.email = email; }

    public String getTelefone() { return telefone; }
    public void setTelefone(String telefone) { this.telefone = telefone; }

    public String getTipo() { return tipo; }
    public void setTipo(String tipo) { this.tipo = tipo; }

    public String getDocumentoFederal() { return documentoFederal; }
    public void setDocumentoFederal(String documentoFederal) { this.documentoFederal = documentoFederal; }

    public String getFotoPerfil() { return fotoPerfil; }
    public void setFotoPerfil(String fotoPerfil) { this.fotoPerfil = fotoPerfil; }

    public String getSenha() { return senha; }
    public void setSenha(String senha) { this.senha = senha; }

    public Double getScore() { return score; }
    public void setScore(Double score) { this.score = score; }

    public Double getMediaAvaliacao() { return mediaAvaliacao; }
    public void setMediaAvaliacao(Double mediaAvaliacao) { this.mediaAvaliacao = mediaAvaliacao; }

    public LocalDate getDataNascimento() { return dataNascimento; }
    public void setDataNascimento(LocalDate dataNascimento) { this.dataNascimento = dataNascimento; }

    public String getCidade() { return cidade; }
    public void setCidade(String cidade) { this.cidade = cidade; }

    public String getEstado() { return estado; }
    public void setEstado(String estado) { this.estado = estado; }

    public String getCnhNumero() { return cnhNumero; }
    public void setCnhNumero(String cnhNumero) { this.cnhNumero = cnhNumero; }

    public String getCnhCategoria() { return cnhCategoria; }
    public void setCnhCategoria(String cnhCategoria) { this.cnhCategoria = cnhCategoria; }

    public LocalDate getCnhValidade() { return cnhValidade; }
    public void setCnhValidade(LocalDate cnhValidade) { this.cnhValidade = cnhValidade; }

    public String getVeiculoModelo() { return veiculoModelo; }
    public void setVeiculoModelo(String veiculoModelo) { this.veiculoModelo = veiculoModelo; }

    public String getVeiculoPlaca() { return veiculoPlaca; }
    public void setVeiculoPlaca(String veiculoPlaca) { this.veiculoPlaca = veiculoPlaca; }

    public Integer getVeiculoAno() { return veiculoAno; }
    public void setVeiculoAno(Integer veiculoAno) { this.veiculoAno = veiculoAno; }

    public String getVeiculoCor() { return veiculoCor; }
    public void setVeiculoCor(String veiculoCor) { this.veiculoCor = veiculoCor; }

    public String getNomeFantasia() { return nomeFantasia; }
    public void setNomeFantasia(String nomeFantasia) { this.nomeFantasia = nomeFantasia; }

    public String getEnderecoComercial() { return enderecoComercial; }
    public void setEnderecoComercial(String enderecoComercial) { this.enderecoComercial = enderecoComercial; }

    public int getTentativasLogin() { return tentativasLogin == null ? 0 : tentativasLogin; }
    public void setTentativasLogin(Integer tentativasLogin) { this.tentativasLogin = tentativasLogin; }

    public LocalDateTime getBloqueadoAte() { return bloqueadoAte; }
    public void setBloqueadoAte(LocalDateTime bloqueadoAte) { this.bloqueadoAte = bloqueadoAte; }

    public LocalDateTime getCriadoEm() { return criadoEm; }
}
